snowshu/adapters/source_adapters/base_source_adapter.py

```diff
@@ -25,16 +42,77 @@ def __init__(self, preserve_case:bool=False):
                 raise NotImplementedError(
                     f'Source adapter requires attribute f{attr} but was not set.')
 
-    def get_all_databases(self) -> Tuple:
-        logger.debug('Collecting databases from snowflake...')
-        databases = tuple(self._safe_query(self.get_all_databases_statement())[
-                          'database_name'].tolist())
-        logger.debug(f'Done. Found {len(databases)} databases.')
-        return databases
+    def build_catalog(self, patterns: Iterable[dict], thread_workers: int = 1) -> Tuple[Relation]:
+        """ This function is expected to return all of the relations that satisfy the filters 
+
+            Args:
+                patterns (Iterable[dict]): Filter dictionaries to apply to the source databases
+                    requires "database", "schema", and "name" keys
+                thread_workers (int): The number of workers to use when building the catalog
+
+            Returns:
+                Tuple[Relation]: All of the relations from the source adapter pass the filters
+        """
+        filtered_schemas = self._get_filtered_schemas(patterns)
+
+        def accumulate_relations(schema_obj: BaseSourceAdapter._DatabaseObject, accumulator):
+            try:
+                relations = self._get_relations_from_database(schema_obj)
+                accumulator += [r for r in relations if at_least_one_full_pattern_match(r, patterns)]
+            except Exception as e:
+                logger.critical(e)
+                raise e
+
+        # get all columns for filtered db/schema
+        catalog = []
+        logger.info('Building filtered catalog...')
+        start_time = time.time()
+        with ThreadPoolExecutor(max_workers=thread_workers) as executor:
+            {executor.submit(accumulate_relations, d, catalog)
+            for d in filtered_schemas}
+
+        logger.info(
+            f'Done building catalog. Found a total of {len(catalog)} relations from the source in {duration(start_time)}.')
```

Comment:
`start_time` will not show the duration just when threads kicked off.

Reply:
That `duration` helper function handles both the difference calculation and the formatting. If we want to avoid that, I can create a new one and essentially deprecate the old one.

Reply:
I think it makes sense - `duration(start_time)` assumes you want the duration between start_time and now si?

Reply:
Yeah. It calculates the difference between now and `start_time` within the helper function.